File: gorg64_spkplay_evdev.c
//    Program for playing melodyes on PC-Speaker.
//    For GNU/Linux 64 bit version. Evdev version.
//    Version: 3.
//    Written on FreePascal (https://freepascal.org/).
//    Copyright (C) 2021-2024  Artyomov Alexander
//    http://self-made-free.ru/
//    aralni@mail.ru

//    This program is free software: you can redistribute it and/or modify
//    it under the terms of the GNU Affero General Public License as
//    published by the Free Software Foundation, either version 3 of the
//    License, or (at your option) any later version.

//    This program is distributed in the hope that it will be useful,
//    but WITHOUT ANY WARRANTY; without even the implied warranty of
//    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//    GNU Affero General Public License for more details.

//    You should have received a copy of the GNU Affero General Public License
//    along with this program.  If not, see <https://www.gnu.org/licenses/>.

#define _GNU_SOURCE
#include <sys/io.h>
#include <unistd.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <fcntl.h>
#include <sys/mman.h>
#include <sys/stat.h>
#include <glob.h>
#include <stdint.h>
#include <linux/input.h>

// gcc -O4 -fPIE -pie -msse4a -ffast-math gorg64_spkplay_evdev.c  -o gorg64_spkplay -lm

int efd;

typedef struct {
	unsigned short tone;
	unsigned short duration;
} ttw;

short spkf(float tone)
{
int64_t tmp;
if ( tone < 1 ) { return 0; };
tmp = round(1193182 / tone);
if (tmp > 0xFFFF) { tmp = 0xFFFF; };
return tmp;
}

int open_evdev(const char *const device_name)
{
    struct stat sb;

    if (-1 == stat(device_name, &sb)) {
        return -1;
    }

    if (!S_ISCHR(sb.st_mode)) {
        return -1;
    }

    const int fd = open(device_name, O_WRONLY);
    if (fd == -1) {
        return -1;
    }

    if (-1 == fstat(fd, &sb)) {
        return -1;
    }

    if (!S_ISCHR(sb.st_mode)) {
        return -1;
    }

    return fd;
}

static
void espk(const uint16_t freq)
{
    struct input_event e;

    memset(&e, 0, sizeof(e));
    e.type = EV_SND;
    e.code = SND_TONE;
    e.value = spkf(freq);

    if (sizeof(e) != write(efd, &e, sizeof(e))) {
        /* If we cannot use the sound API, we cannot silence the sound either */
        _Exit(1);
    }
}

static
void espkoff()
{
    struct input_event e;

    memset(&e, 0, sizeof(e));
    e.type = EV_SND;
    e.code = SND_TONE;
    e.value = 0;

    if (sizeof(e) != write(efd, &e, sizeof(e))) {
        _Exit(1);
    }
}

void SIGHUPHandler(int signal) {
puts("SIGHUP");
espkoff();
_Exit(0);
}
void SIGINTHandler(int signal) {
puts("SIGINT");
espkoff();
_Exit(0);
}
void SIGTERMHandler(int signal) {
puts("SIGTERM");
espkoff();
_Exit(0);
}
void InstallSignalHandlers() {
  struct sigaction action, nu;
  memset(&action, 0, sizeof(action));
  memset(&nu, 0, sizeof(nu));
  action.sa_handler = SIGTERMHandler;
  action.sa_flags = SA_RESTART;
  sigaction(SIGTERM, &action, &nu);
  action.sa_handler = SIGINTHandler;
  sigaction(SIGINT, &action, &nu);
  action.sa_handler = SIGHUPHandler;
  sigaction(SIGHUP, &action, &nu);
}

int
main(int argc, char *argv[])
{
long int pid = getpid();
glob_t globlist;
int i;
char cm[1024];
struct stat sp;
i = 0;               
glob("/proc/[0-9]*", GLOB_ONLYDIR, NULL, &globlist);
      while (globlist.gl_pathv[i])
        {
char *name = strrchr(globlist.gl_pathv[i], '/') + 1;
if (pid != atoi(name)) {
FILE *fp = fopen(strcat(globlist.gl_pathv[i], "/comm"), "r");
fgets(cm, 1023, fp);
cm[strcspn(cm, "\n" )] = '\0';
fclose(fp);
if ((strcmp("gorg64_spkplay", cm) == 0) | (strcmp("gorg64_spktone", cm) == 0)) {
puts("Already running. Exit.");
_Exit(0);
}
}
          i++;
        }
globfree(&globlist);

InstallSignalHandlers();

efd = open_evdev("/dev/input/by-path/platform-pcspkr-event-spkr");
if (efd < 0) {puts("Error open evdev"); _Exit(1);} else {puts("Use evdev");}

if (argc == 1) {puts("Use: gorg64_spkplay file1.speaker file2.speaker ... ");
espk(1000); usleep(1000000); espkoff(); _Exit(0);}

void *a;
ttw *p;
struct stat sb;
int fd;
int64_t ff;
int64_t f;
for (ff = 1;  ff < argc; ff++){
puts(argv[ff]);
fd = open(argv[ff], O_RDONLY);
fstat(fd, &sb);
a = mmap(NULL,  sb.st_size, PROT_READ, MAP_PRIVATE, fd,  0);
close(fd);
if (a == MAP_FAILED) {puts("map failed"); continue;}
if (sb.st_size % 4 != 0) {puts("size / 4 <> 0"); continue;}
p = a;
for (f = 0; f < sb.st_size; f++){
 if (p->duration < 1) continue;
 if (p->tone < 1) {
   espkoff();
   usleep(p->duration*1000);
 } else {
   espk(p->tone);
   usleep(p->duration*1000);
 }
p++;
}
espkoff();

munmap(a, sb.st_size);
}
_Exit(0);
}

File: gorg64_spkplay_evdev.pas
program gorg64_spkplay_evdev;

{$MODE OBJFPC}
{$RANGECHECKS ON}
{$LONGSTRINGS ON}
{$SMARTLINK ON}
{$ASMMODE INTEL}

{
    Program for playing melodys on PC-Speaker.
    For GNU/Linux 64 bit version.
    Version: 4.
    Written on FreePascal (https://freepascal.org/).
    Copyright (C) 2021,2022,2024  Artyomov Alexander
    http://self-made-free.ru/
    aralni@mail.ru

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU Affero General Public License as
    published by the Free Software Foundation, either version 3 of the
    License, or (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Affero General Public License for more details.

    You should have received a copy of the GNU Affero General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
}

uses sysutils,unix,baseunix,linux,urun;

var
  efd : Integer;

const
EV_SND = 18;
SND_TONE = 2;

type
      TTW = packed record
       tone, duration : Word;
      end;
      PTW=^TTW;

var
    f, ff : Int64;
    oa,na : PSigActionRec;
    fs : Int64 = 0;
    a :  PTW = nil;
    fFileName : utf8string = '';

function spkf(tone : Word) : Word;
var
  tmp : Int64;
begin
if tone < 1 then Exit(0);
tmp := 1193182 div tone;
if tmp > $FFFF then tmp := $FFFF;
Exit(tmp);
end;
function spkf(tone : extended) : Word;
var
  tmp : Int64;
begin
if tone < 1 then Exit(0);
tmp := round(1193280 / tone);
if tmp > $FFFF then tmp := $FFFF;
Exit(tmp);
end;

function open_evdev(fn : string) : Integer;
var
 info : stat;
 fd : cint;
begin
if fpstat (fn,info)<>0 then
begin
	writeln('Fstat failed. Errno : ',fpgeterrno);
	halt (1);
end;
if not fpS_ISCHR(info.st_mode) then begin
	writeln('fpS_ISCHR failed.');
	halt (1);
end;
FD:=fpOpen(fn,O_WrOnly);
if not(FD>0) then
begin
	writeln('fpOpen failed.');
	halt (1);
end;
if fpfstat (fd,info)<>0 then
begin
	writeln('Fstat 2 failed. Errno : ',fpgeterrno);
	halt (1);
end;
if not fpS_ISCHR(info.st_mode) then begin
	writeln('fpS_ISCHR failed.');
	halt (1);
end;
Exit(fd);
end;

type
input_event = record
	time : timeval;
	_type : Word;
	code : Word;
	value : LongInt;
end;

procedure espk(freq : Word);
var
    e : input_event;
begin
//    memset(&e, 0, sizeof(e));
    e._type := EV_SND;
    e.code := SND_TONE;
    e.value := spkf(freq);

    if (sizeof(e) <> fpwrite(efd, e, sizeof(e))) then begin
//        /* If we cannot use the sound API, we cannot silence the sound either */
        Halt(1);
    end;
end;

procedure espkoff;
var
    e : input_event;
begin
//    memset(&e, 0, sizeof(e));
    e._type := EV_SND;
    e.code := SND_TONE;
    e.value := 0;

    if (sizeof(e) <> fpwrite(efd, e, sizeof(e))) then begin
        Halt(1);
    end;
end;

function LoadFromFile(fn : utf8string) : boolean;
var
  fp : File of TTW;
begin
Assign(fp, fn);
FileMode := 0;
{$I-}
ReSet(fp);
{$I+} if IOResult <> 0 then Exit(true);
{$I-}
fs := FileSize(fp);
{$I+} if IOResult <> 0 then Exit(true);
GetMem(a, fs*SizeOf(TTW));
{$I-}
BlockRead(fp, a[0], fs);
{$I+} if IOResult <> 0 then Exit(true);
{$I-}
Close(fp);
{$I+} if IOResult <> 0 then Exit(true);
fFileName := fn;
Exit(false);
end;

Procedure DoSig(sig : cint);cdecl;
begin
   writeln('Receiving signal: ',sig);
   espkoff;
   halt(0);
end;

begin
WriteLn('GALAXY ORGANIZER SPEAKER PLAYER Version 3');
WriteLn('Artyomov Alexander 2022-2024  License: GNU AGPLv3 and above');
WriteLn('Use: gorg64_spkplay_syscalls or gorg64_spkplay_syscalls somemusic.speaker somemusic2.speaker ...');

efd := open_evdev('/dev/input/by-path/platform-pcspkr-event-spkr');
//WriteLn('EFD = ',efd);

   new(na);
   new(oa);
   na^.sa_Handler:=SigActionHandler(@DoSig);
   fillchar(na^.Sa_Mask,sizeof(na^.sa_mask),#0);
   na^.Sa_Flags:= SA_RESTART;
   na^.Sa_Restorer:=Nil;
   if fpSigAction(SigTerm,na,oa)<>0 then
     begin
     writeln('Error: ',fpgeterrno,'.');
     halt(1);
     end;
   if fpSigAction(SigHup,na,oa)<>0 then
     begin
     writeln('Error: ',fpgeterrno,'.');
     halt(1);
     end;
   if fpSigAction(SigInt,na,oa)<>0 then
     begin
     writeln('Error: ',fpgeterrno,'.');
     halt(1);
     end;
   if fpSigAction(SigQuit,na,oa)<>0 then
     begin
     writeln('Error: ',fpgeterrno,'.');
     halt(1);
     end;
   if fpSigAction(SigTStp,na,oa)<>0 then
     begin
     writeln('Error: ',fpgeterrno,'.');
     halt(1);
     end;

if ParamCount = 0 then begin
 espk(1000); sleep(1000); espkoff;
 Halt;
end;

for ff := 1 to ParamCount do begin
if LoadFromFile(ParamStr(ff)) then begin
 WriteLn('Err');  espk(300); sleep(1000); espkoff;
Halt(2);
end;
WriteLn('* Playing file: ' + fFileName);

for f := 0 to fs-1 do begin
 if a[f].duration < 1 then continue;
 if a[f].tone < 1 then begin
   espkoff;
   sleep(a[f].duration);
 end else begin
   espk(a[f].tone);
   sleep(a[f].duration);
 end;
end;
espkoff;
FreeMem(a);
end;

end.